File: scripts/download_usc_zip.py
import os
import requests
from bs4 import BeautifulSoup
from zipfile import ZipFile
from tqdm import tqdm

RAW_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '../data/raw'))
XML_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '../data/xml'))
ZIP_OUTPUT_PATH = os.path.join(RAW_DIR, 'usc.zip')
DOWNLOAD_PAGE_URL = 'https://uscode.house.gov/download/download.shtml'


def ensure_dirs():
    os.makedirs(RAW_DIR, exist_ok=True)
    os.makedirs(XML_DIR, exist_ok=True)

def get_bulk_xml_url():
    print('Fetching download page...')
    resp = requests.get(DOWNLOAD_PAGE_URL)
    resp.raise_for_status()
    soup = BeautifulSoup(resp.text, 'html.parser')
    link = soup.find('a', title='All USC Titles in XML')
    if not link or not link.get('href'):
        print('Could not find <a> tag with title="All USC Titles in XML". Dumping a snippet of HTML:')
        print(resp.text[:2000])
        raise RuntimeError('Could not find XML ZIP link on download page')
    href = link['href']
    url = href if href.startswith('http') else f'https://uscode.house.gov{href}'
    print('Found bulk XML ZIP URL:', url)
    return url

def download_file(url, output_path):
    print(f'Downloading {url} to {output_path}...')
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        total = int(r.headers.get('content-length', 0))
        with open(output_path, 'wb') as f, tqdm(total=total, unit='B', unit_scale=True, desc='Downloading ZIP') as pbar:
            for chunk in r.iter_content(chunk_size=8192):
                if chunk:
                    f.write(chunk)
                    pbar.update(len(chunk))
    print('Download completed successfully.')

def unzip_to_xml_dir(zip_path, out_dir):
    print(f'Unzipping {zip_path} to {out_dir}...')
    with ZipFile(zip_path, 'r') as zip_ref:
        zip_ref.extractall(out_dir)
    print('Unzip complete.')

def main():
    ensure_dirs()
    url = get_bulk_xml_url()
    download_file(url, ZIP_OUTPUT_PATH)
    unzip_to_xml_dir(ZIP_OUTPUT_PATH, XML_DIR)
    print('USC XML download and extraction complete.')

if __name__ == '__main__':
    main()

File: scripts/download_usc_zip.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { createWriteStream, createReadStream } from 'fs';
import puppeteer from 'puppeteer';
import https from 'https';
import unzipper from 'unzipper';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DOWNLOAD_PAGE_URL = 'https://uscode.house.gov/download/download.shtml';
const OUTPUT_DIR = path.resolve('data/raw');
const ZIP_OUTPUT_PATH = path.join(OUTPUT_DIR, 'usc.zip');

async function getBulkXmlUrlWithPuppeteer() {
  const browser = await puppeteer.launch({ headless: true });
  const page = await browser.newPage();
  await page.goto(DOWNLOAD_PAGE_URL, { waitUntil: 'domcontentloaded' });
  // Find the link in the page context
  const url = await page.evaluate(() => {
    const link = document.querySelector('a[title="All USC Titles in XML"]');
    return link ? link.href : null;
  });
  await browser.close();
  if (!url) throw new Error('Could not find XML ZIP link on download page');
  return url;
}

async function downloadZip(url, outputPath) {
  await fs.mkdir(path.dirname(outputPath), { recursive: true });
  return new Promise((resolve, reject) => {
    const file = createWriteStream(outputPath);
    https.get(url, response => {
      if (response.statusCode !== 200) {
        reject(new Error(`Failed to download: ${response.statusCode}`));
        return;
      }
      response.pipe(file);
      file.on('finish', async () => {
        file.close();
        // Validate ZIP file
        try {
          const fd = await fs.open(outputPath, 'r');
          const buffer = Buffer.alloc(4);
          await fd.read(buffer, 0, 4, 0);
          await fd.close();
          const isZip = buffer.slice(0, 2).toString() === 'PK';
          if (!isZip) {
            const first200 = (await fs.readFile(outputPath)).toString('utf8', 0, 200);
            reject(new Error('Downloaded file is not a valid ZIP archive. First 200 bytes: ' + first200));
            return;
          }
          // Extract ZIP to data/xml
          const extractDir = path.resolve('data/xml');
          await fs.mkdir(extractDir, { recursive: true });
          await new Promise((resolveExtract, rejectExtract) => {
            createReadStream(outputPath)
              .pipe(unzipper.Extract({ path: extractDir }))
              .on('close', resolveExtract)
              .on('error', rejectExtract);
          });
          console.log('Extraction completed to', extractDir);
          resolve();
        } catch (err) {
          reject(new Error('Failed to validate ZIP file: ' + err));
        }
      });
    }).on('error', err => {
      fs.unlink(outputPath);
      reject(err);
    });
  });
}

async function main() {
  try {
    console.log('Launching headless browser to fetch download page...');
    const zipUrl = await getBulkXmlUrlWithPuppeteer();
    console.log('Found ZIP URL:', zipUrl);
    console.log(`Downloading ZIP to ${ZIP_OUTPUT_PATH}...`);
    await downloadZip(zipUrl, ZIP_OUTPUT_PATH);
    console.log('Download completed successfully.');
  } catch (err) {
    console.error('Error:', err);
    process.exit(1);
  }
}

main();
